### PR TITLE
add option for multiple application configurations

### PR DESCRIPTION
**Pull request by lievenhey**
I added an option to create multiple configurations for one application. This way you can quickly test an application in different configurations.

Stuff that needs to be done:
- improve UI (looks ugly but I am not sure how to make it look better)
- check localization

### CODE REVIEW

File: src/recordpage.cpp [hunk 1]
```diff
@@ -207,6 +207,21 @@ RecordPage::RecordPage(QWidget* parent)
     ui->outputFile->setMode(KFile::File | KFile::LocalOnly);
     ui->eventTypeBox->lineEdit()->setPlaceholderText(tr("perf defaults (usually cycles:Pu)"));
 
+    ui->launchConfigComboBox->setEditable(true);
+    ui->launchConfigComboBox->setDisabled(true);
+    ui->launchConfigComboBox->setInsertPolicy(QComboBox::InsertAtCurrent);
+    connect(ui->launchConfigComboBox->lineEdit(), &QLineEdit::editingFinished, this, &RecordPage::onApplicationConfigRenamed);
+    connect(ui->launchConfigComboBox, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &RecordPage::onApplicationConfigChanged);
```
Comment (milianw) [inline src/recordpage.cpp:214]: use qOverload<int>(&...)

File: src/recordpage.cpp [hunk 2]
```diff
@@ -378,7 +393,7 @@ RecordPage::RecordPage(QWidget* parent)
     connect(stopRecordingShortcut, &QShortcut::activated, this,
             [this] { ui->startRecordingButton->setChecked(false); });
 
-    auto* startRecordingShortcut = new QShortcut(Qt::Key_Return, this);
+    auto* startRecordingShortcut = new QShortcut(Qt::CTRL + Qt::Key_Return, this);
```
Comment (milianw) [inline src/recordpage.cpp:396]: this change of the shortcut should be a separate commit with an explanation of why you change it

File: src/recordpage.cpp [hunk 3]
```diff
@@ -605,10 +620,12 @@ void RecordPage::onApplicationNameChanged(const QString& filePath)
         setError(tr("Application file is not executable: %1").arg(filePath));
     } else {
         const auto config = applicationConfig(filePath);
-        ui->workingDirectory->setText(config.readEntry("workingDir", QString()));
+        ui->workingDirectory->setText(config.readEntry("workdir", QString()));
```
Comment (milianw) [inline src/recordpage.cpp:623]: why did you change the key here, this could stay at the old name, no?

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/recordpage.cpp (hunk 3 above)
Comment (lievenhey) [reply src/recordpage.cpp:623]: I changed this one by mistage

File: src/recordpage.cpp [hunk 4]
```diff
@@ -664,6 +681,43 @@ void RecordPage::onOutputFileNameSelected(const QString& filePath)
...
+    const int index = ui->launchConfigComboBox->currentIndex();
+
+    const auto oldText = ui->launchConfigComboBox->itemData(index).toString();
+
+    group.deleteGroup(ui->launchConfigComboBox->itemData(index).toString());
+
+    const auto currentText = ui->launchConfigComboBox->currentText();
+    saveApplicationConfig(currentText);
+
+    ui->launchConfigComboBox->setItemData(index, currentText);
```
[the remark this answers is not in the document]
Comment (lievenhey) [reply src/recordpage.cpp:704]: I use the ItemData to identify the old configuration after the name got changed. Here I need to update to old name to the new one.

File: src/recordpage.cpp [hunk 5]
```diff
@@ -207,6 +207,21 @@ RecordPage::RecordPage(QWidget* parent)
...
 
+    ui->launchConfigComboBox->setEditable(true);
+    ui->launchConfigComboBox->setDisabled(true);
+    ui->launchConfigComboBox->setInsertPolicy(QComboBox::InsertAtCurrent);
+    connect(ui->launchConfigComboBox->lineEdit(), &QLineEdit::editingFinished, this, &RecordPage::onApplicationConfigRenamed);
+    connect(ui->launchConfigComboBox, QOverload<int>::of(&QComboBox::currentIndexChanged), this, &RecordPage::onApplicationConfigChanged);
+    connect(ui->addConfig, &QPushButton::clicked, this, &RecordPage::onApplicationConfigAdded);
+    connect(ui->removeConfig, &QPushButton::clicked, this, &RecordPage::onApplicationConfigRemoved);
+
+    const auto updateLaunchConfig = [this]{if (ui->launchConfigComboBox->count() > 0) {
```
[the remark this answers is not in the document]
Comment (lievenhey) [reply src/recordpage.cpp:219]: done

[the remark this answers is not in the document]
Comment (milianw) [reply src/recordpage.cpp]: still looks odd

File: src/recordpage.cpp [hunk 6]
```diff
@@ -664,6 +682,42 @@ void RecordPage::onOutputFileNameSelected(const QString& filePath)
     }
 }
 
+void RecordPage::onApplicationConfigAdded()
+{
+    saveApplicationConfig(QLatin1String("Config ") + QString::number(ui->launchConfigComboBox->count() + 1));
```
Comment (milianw) [inline src/recordpage.cpp:689]: QLatin1String("Config %1").arg()

File: src/recordpage.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (milianw) [reply src/recordpage.cpp:706]: ah I see. that's a creative idea but you should totally document this in a comment somewhere as it's not self-evident

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/recordpage.cpp [hunk 7]
```diff
@@ -664,6 +685,44 @@ void RecordPage::onOutputFileNameSelected(const QString& filePath)
     }
 }
 
+void RecordPage::onApplicationConfigAdded()
+{
+    saveApplicationConfig(QLatin1String("Config %1").arg(QString::number(ui->launchConfigComboBox->count() + 1)));
```
Comment (milianw) [inline src/recordpage.cpp:690]: QString::number isn't needed anymore now

File: src/recordpage.cpp [hunk 8]
```diff
@@ -689,6 +748,56 @@ void RecordPage::updateProcessesFinished()
...
+void RecordPage::saveApplicationConfig(const QString& name)
+{
+    const auto& group = applicationConfig(ui->applicationName->text());
+    KConfigGroup config(&group, name);
+    config.writeEntry("params", ui->applicationParametersBox->text());
+    config.writeEntry("workingDir", ui->workingDirectory->text());
+    config.sync();
+}
+
+void RecordPage::applyApplicationConfig(const QString name)
```
Comment (milianw) [inline src/recordpage.cpp:760]: const&

File: src/recordpage.cpp [hunk 9]
```diff
@@ -689,6 +748,56 @@ void RecordPage::updateProcessesFinished()
...
+{
+    if (name.isEmpty())
+        return;
+    const auto& config = applicationConfig(ui->applicationName->text()).group(name);
+    ui->applicationParametersBox->setText(config.readEntry("params", ""));
+
```
Comment (milianw) [inline src/recordpage.cpp:766]: move empty line two up:

```
if (name.isEmpty())
    return;

const auto& config = ...;
...->setText();
...->setText();
```

File: src/recordpage.cpp [hunk 10]
```diff
@@ -664,6 +685,44 @@ void RecordPage::onOutputFileNameSelected(const QString& filePath)
...
+{
+    auto appConfig = applicationConfig(ui->applicationName->text());
+    appConfig.deleteGroup(ui->launchConfigComboBox->currentText());
+
+    updateListOfApplicationConfigs();
+}
+
+void RecordPage::onApplicationConfigChanged(int index)
+{
+    Q_UNUSED(index);
```
Comment (milianw) [inline src/recordpage.cpp:722]: just remove the index? and I think such trivial one-liners could just be a lambda directly connected to the signal

File: src/recordpage.cpp (hunk 7 above)
Comment (lievenhey) [reply src/recordpage.cpp]: it is needed. int does not convert implicit to QLatin1String.

Comment (lievenhey) [reply src/recordpage.cpp]: I can change it to QString, since arg on a QString does support int

File: src/recordpage.cpp [hunk 11]
```diff
@@ -664,6 +685,50 @@ void RecordPage::onOutputFileNameSelected(const QString& filePath)
...
+void RecordPage::onApplicationConfigRemoved()
+{
+    auto appConfig = applicationConfig(ui->applicationName->text());
+    appConfig.deleteGroup(ui->launchConfigComboBox->currentText());
+
+    updateListOfApplicationConfigs();
+
+    if (ui->launchConfigComboBox->count() == 0) {
+        // restore default configuration values
+        const auto config = applicationConfig(ui->applicationName->text());
```
Comment (milianw) [inline src/recordpage.cpp:721]: reuse code:

```
applyApplicationConfig(ui->applicationName->text());
```

**Review by milianw — CHANGES_REQUESTED** [review]

**Comment by milianw** [comment]
thanks!
